lib/fs-writable.js

```diff
@@ -0,0 +1,160 @@
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+
+'use strict';
+
+const fs = require('fs');
+
+const Writable = require('./writable');
+const {
+  ERR_INVALID_ARG_TYPE,
+  ERR_OUT_OF_RANGE
+} = require('./errors');
+
+const openSymbol = Symbol('open');
+const closeSymbol = Symbol('close');
+const errorSymbol = Symbol('error');
+const closeStreamSymbol = Symbol('closeStream');
+
+class WritableFileStream extends Writable {
+  constructor(path, {
+    fd = null,
+    start,
+    flags = 'w',
+    bufferSize,
+    mode = 0o666,
+    encoding = null,
+    autoClose = true,
+  } = {}) {
+    super({ bufferSize });
+
+    this.fd = fd;
+    this.path = path;
+
+    this.closed = false;
+    this.bytesWritten = 0;
+
+    this.autoClose = autoClose;
+
+    if (typeof path !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('path', 'string', path);
+    }
+
+    if (start !== undefined) {
+      if (typeof start !== 'number') {
+        throw new ERR_INVALID_ARG_TYPE('start', 'number', start);
+      }
+      if (start < 0) {
+        throw new ERR_OUT_OF_RANGE('start', '>= 0', `{start: ${start}}`);
+      }
+
+      this.pos = start;
+    }
+
+    if (encoding) this.setDefaultEncoding(encoding);
+
+    if (this.fd === null) this[openSymbol](flags, mode);
+  }
+
+  [errorSymbol](error, cb) {
+    if (this.autoClose) {
+      this.destroy(error);
+    } else {
+      this.emit('error', error);
+      this.emit('close');
+    }
+    if (cb) cb(error);
+  }
+
+  [openSymbol](flags, mode) {
+    fs.open(this.path, flags, mode, (error, fd) => {
+      if (error) {
+        this[errorSymbol](error);
+        return;
+      }
+
+      this.fd = fd;
+      this.emit('open', fd);
+    });
+  }
+
+  [closeSymbol](cb) {
+    if (cb) {
+      if (this.closed) {
+        process.nextTick(cb);
+        return;
+      } else {
+        this.once('close', cb);
+      }
+    }
+
+    if (!this.autoClose) {
+      this.once('finish', this.destroy.bind(this));
+    }
+
+    this.end();
+  }
+
+  [closeStreamSymbol](cb, err) {
+    fs.close(this.fd, (er) => {
+      er = er || err;
+      cb(er);
+      this.closed = true;
+      if (!er) this.emit('close');
```

Comment:
What's the reason behind not emitting 'close' here but still emitting it in case of error in `[errorSymbol]`?